java/src/main/java/cucumber/runtime/java/MethodScanner.java

```diff
@@ -56,26 +53,25 @@ public void scan(JavaBackend javaBackend, List<String> gluePaths) {
      * @param glueCodeClass the class where the method is declared.
      */
     public void scan(JavaBackend javaBackend, Method method, Class<?> glueCodeClass) {
-        for (Class<? extends Annotation> cucumberAnnotationClass : cucumberAnnotationClasses) {
-            Annotation annotation = method.getAnnotation(cucumberAnnotationClass);
-            if (annotation != null) {
-                if (!method.getDeclaringClass().isAssignableFrom(glueCodeClass)) {
-                    throw new CucumberException(String.format("%s isn't assignable from %s", method.getDeclaringClass(), glueCodeClass));
-                }
-                if (!glueCodeClass.equals(method.getDeclaringClass())) {
-                    throw new CucumberException(String.format("You're not allowed to extend classes that define Step Definitions or hooks. %s extends %s", glueCodeClass, method.getDeclaringClass()));
-                }
-                if (isHookAnnotation(annotation)) {
-                    javaBackend.addHook(annotation, method);
-                } else if (isStepdefAnnotation(annotation)) {
-                    javaBackend.addStepDefinition(annotation, method);
-                }
+        Annotation[] methodAnnotations = method.getAnnotations();
+        for (Annotation annotation : methodAnnotations) {
```

Comment:
Might be even more efficient to use Method.getAnnotation(Class<T> annotationClass) rather then checking all annotations. Then we can replace this for loop with two null checks.

Reply:
I'm not sure about it. Could you clarify? Based on previous logic one method can be annotated with Before/After and any number of step annotations  - we can't achieve it without loop. Second thing is that we are not able to get step annotation e.g. `@Given` using `Method.getAnnotation(Class annotationClass)` because we don't know what annotation is used (actually cucumber does not care if it is `@Given` or `@Then` it just should be annotation that itself is annotated with `@StepDefAnnotation` ), eventually this code won't work: `method.getAnnotation(StepDefAnnotation.class)`

Reply:
Derp. You are correct. I didn't have hierarchy of annotations properly in my mind.